### PR TITLE
[recorder] Switch OMXCodec for MediaCodecSource for video recording. Fixes JB#39030

### PR DESCRIPTION
**Pull request by abranson**
Originally required another Android service, but I made a fake one instead.

### CODE REVIEW

**Comment by abranson** [comment]
This seems to work fine on Android 6 (32 and 64bit) and Android 5 (32 bit). It doesn't seem to work on community xperia builds, but I think that's an issue to be tackled separately.

File: droidmediacodec.cpp [hunk 1]
```diff
@@ -17,14 +17,26 @@
...
+#include <media/stagefright/MediaCodecSource.h>
+#include <media/stagefright/MediaCodec.h>
+#include <media/openmax/OMX_Video.h>
+#include <media/openmax/OMX_Audio.h>
+#include <media/stagefright/ACodec.h>
+#include <media/stagefright/foundation/ALooper.h>
+#include <media/stagefright/foundation/AMessage.h>
+#endif
+
 #if ANDROID_MAJOR < 7
```
Comment (saukko) [inline droidmediacodec.cpp]: should this be in #else of the above instead? This is only for ANDROID_MAJOR < 5 now right?

File: droidmediacodec.cpp [hunk 2]
```diff
@@ -317,17 +335,95 @@ class DroidMediaCodecBuilder {
...
-#if ANDROID_MAJOR < 7
-    return android::OMXCodec::Create(omx.interface(),
-				     md,
-				     isEncoder(),
-				     src,
-				     NULL, flags(), window);
+
+#if ANDROID_MAJOR >= 5
+  const char *mime;
+  md->findCString(android::kKeyMIMEType, &mime);
```
Comment (saukko) [inline droidmediacodec.cpp]: Identation seems wrong here.

File: droidmediacodec.cpp (hunk 1 above)
Comment (abranson) [reply droidmediacodec.cpp]: No that's for 6 and below. There are constants in the flags that only got moved in 7 (kSoftwareCodecsOnly & kHardwareCodecsOnly).

File: droidmediacodec.cpp (hunk 2 above)
Comment (abranson) [reply droidmediacodec.cpp]: Fixed

File: droidmediacodec.cpp [hunk 3]
```diff
@@ -317,17 +335,95 @@ class DroidMediaCodecBuilder {
...
+
+    if (isEncoder()) {
+      if (!strncmp("video", mime, 5)) {
+      android::sp<android::AMessage> format = new android::AMessage();
+
+      format->setString("mime", mime);
+      ALOGW("Creating video encoder for %s", mime);
+      int32_t width, height, stride, sliceh, colour, frames, maxinput, bitrate, iframes;
+
+      md->findInt32(android::kKeyWidth, &width);
```
Comment (saukko) [inline droidmediacodec.cpp]: I wonder if we should have if (md->findInt32()) format->setInt32()? Do we need to check if width > 0 here?

Comment (abranson) [reply droidmediacodec.cpp]: Yes I think that would be simpler. Will change.

File: droidmediacodec.cpp [hunk 4]
```diff
@@ -317,17 +335,109 @@ class DroidMediaCodecBuilder {
...
-     return src;
+      return android::MediaCodecSource::Create(looper, format, src, flags());
 #endif
+      }
+      else {
+        android::sp<android::AMessage> format = new android::AMessage();
+        format->setString("mime", mime);
+        ALOGW("Creating audio encoder for %s", mime);
+        format->setInt32("aac-profile", OMX_AUDIO_AACObjectLC);
+        int32_t maxinput, samplerate, bitrate;
```
Comment (saukko) [inline droidmediacodec.cpp]: samplerate not used currently.

Comment (abranson) [reply droidmediacodec.cpp]: Removed

**Pull request merged**